contracts/most/UtBase.sol

```diff
@@ -0,0 +1,228 @@
+    uint8 public constant TOKEN_DECIMALS = 18;
+
+
+    /* Storage */
+
+    /** Consensus cogateway address defined in genesis. */
+    address private consensusCogateway = address(0x0000000000000000000000000000000000004d02);
+
+
+    /* Modifiers */
+
+    modifier onlyConsensusCogateway()
+    {
+        require(
+            msg.sender == getConsensusCogateway(),
+            "Caller must be consensus cogateway contract address."
+        );
+        _;
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice Initializes the UtBase contract.
+     *
+     * @dev This function must be called only once.
+     *
+     *      Function requires:
+     *          - msg.sender should be Coconsensus
+     *          - setup can be called only once
+     */
+    function setup()
+        external
+        onlyCoconsensus
+    {
+        require(
+            totalTokenSupply == uint256(0),
+            "Setup is already done."
+        );
+
+        tokenSymbol = TOKEN_SYMBOL;
+        tokenName = TOKEN_NAME;
+        tokenDecimals = TOKEN_DECIMALS;
+        totalTokenSupply = genesisTotalSupply;
+
+        /*
+         * In the genesis block, validators gets base token equivalent to the
+         * amount they deposited before joining the core. Equivalent
+         * amount of ERC20 token should be minted and UtBase contract should
+         * hold that balance. Similar flow will happen in normal mint process
+         * where UtBase contract ERC20 balance will be incremented. So in the
+         * setup, UtBase contract holds the ERC20 token balance equal to the
+         * genesisTotalSupply.
+         */
+        balances[address(this)] = genesisTotalSupply;
+    }
+
+    /**
+     * @notice Mints the base coin equivalent amount to beneficiary address.
+     *
+     * @dev Function requires :
+     *          - Caller must be Consensus cogateway contract.
+     *
+     * @param _beneficiary Beneficairy address which will receive base coin.
+     * @param _amount Amount of UtBase tokens to be minted.
+     *
+     */
+    function mint(
+        address payable _beneficiary,
+        uint256 _amount
+    )
+        external
+        onlyConsensusCogateway
+    {
+        _mint(_beneficiary, _amount);
+        unwrapInternal(_beneficiary, _amount);
+    }
+
+    /**
+     * @notice Unwrap converts ERC20 UtBase to base coin.
+     *
+     * @dev  This contract's base coin balance must always be greater than
+     *       unwrap amount.
+     *
+     * @param _amount Amount of ERC20 UtBase token to convert to base coin.
+     */
+    function unwrap(
+        uint256 _amount
+    )
+        external
+    {
+        unwrapInternal(msg.sender, _amount);
+    }
+
+    /**
+     * @notice Wrap converts base coin to ERC20 UtBase.
+     *
+     * @dev The ERC20 OST balance of contract should always be greater than the
+     *      received payable amount.
+     *
+     *      Function requires:
+     *          - msg.value should be non-zero
```

Comment:
As per the latest developer guidelines, please use pre and post conditions.

Reply:
Done